### PR TITLE
NMS-10721: Respect flow timeout

### PR DESCRIPTION
**Pull request by fooker**
This is another shot in trying to fix flows with timeouts.

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-10721

### CODE REVIEW

File: features/telemetry/protocols/netflow/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/netflow/adapter/ipfix/IpfixFlow.java [hunk 1]
```diff
@@ -420,4 +421,17 @@ public Integer getVlan() {
                 .map(Long::intValue)
                 .orElse(null);
     }
+
+    @Override
+    public Optional<Timeout> getTimeout() {
+        // TODO fooker: Make this pretty
+        final Optional<Long> active = getInt64(this.document, "flowActiveTimeout").map(t -> t * 1000L);
+        final Optional<Long> inactive = getInt64(this.document, "flowActiveTimeout").map(t -> t * 1000L);
```
Comment (christianpape) [inline features/telemetry/protocols/netflow/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/netflow/adapter/ipfix/IpfixFlow.java]: shouldn't this be "flowInactiveTimeout"

Comment (fooker) [reply features/telemetry/protocols/netflow/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/netflow/adapter/ipfix/IpfixFlow.java]: Fixed, thanks.

File: features/telemetry/protocols/netflow/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/netflow/adapter/common/UpdatingFlow.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    public abstract Optional<Timeout> getTimeout();
+
+    @Override
+    public Long getTimeoutSwitched() {
+        // The timespan for this flow export depends on the timeout, if present: use the active or inactive timeout,
+        // whether the flow was active or not, and subtract it from last switched timestamp. Short living flows are
+        // capped to the real first switched timestamp. If there is no timeout information available the first switched
+        // timestamp is used a s an fallback.
```
Comment (Naicisum) [inline features/telemetry/protocols/netflow/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/netflow/adapter/common/UpdatingFlow.java]: Language fix, `used a s an` to `used as an`

File: features/telemetry/protocols/netflow/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/netflow/adapter/netflow9/Netflow9Flow.java [hunk 3]
```diff
@@ -275,6 +276,18 @@ public Integer getVlan() {
...
+    public Optional<Timeout> getTimeout() {
+        final Optional<Long> active = getInt64(this.document, "FLOW_ACTIVE_TIMEOUT").map(t -> t * 1000L);
+        final Optional<Long> inactive = getInt64(this.document, "FLOW_INACTIVE_TIMEOUT").map(t -> t * 1000L);
+
+        if (active.isPresent() && inactive.isPresent()) {
+            return Optional.of(new Timeout(active.get(), inactive.get()));
+        } else {
+            return Optional.empty();
+        }
+    }
```
Comment (Naicisum) [inline features/telemetry/protocols/netflow/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/netflow/adapter/netflow9/Netflow9Flow.java]: Should this entire function move to `UpdatingFlow`?

File: features/telemetry/protocols/netflow/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/netflow/adapter/ipfix/IpfixFlow.java [hunk 4]
```diff
@@ -420,4 +421,16 @@ public Integer getVlan() {
...
+    public Optional<Timeout> getTimeout() {
+        final Optional<Long> active = getInt64(this.document, "flowActiveTimeout").map(t -> t * 1000L);
+        final Optional<Long> inactive = getInt64(this.document, "flowInactiveTimeout").map(t -> t * 1000L);
+
+        if (active.isPresent() && inactive.isPresent()) {
+            return Optional.of(new Timeout(active.get(), inactive.get()));
+        } else {
+            return Optional.empty();
+        }
+    }
```
Comment (Naicisum) [inline features/telemetry/protocols/netflow/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/netflow/adapter/ipfix/IpfixFlow.java]: Should this entire function move to UpdatingFlow?

File: features/telemetry/protocols/netflow/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/netflow/adapter/common/UpdatingFlow.java (hunk 2 above)
Comment (fooker) [reply features/telemetry/protocols/netflow/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/netflow/adapter/common/UpdatingFlow.java]: Fixed

File: features/telemetry/protocols/netflow/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/netflow/adapter/netflow9/Netflow9Flow.java (hunk 3 above)
Comment (fooker) [reply features/telemetry/protocols/netflow/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/netflow/adapter/netflow9/Netflow9Flow.java]: Then we have to extract two functions to return the different key names. I would like to keep it there to have the flexibility to have protocol specific logic.

**Comment by fooker** [comment]
> Can we add an IT for this? I'm thinking we could manually construct the flows using something like the `FlowBuilder`, feed the adapter, persist to ES, and then issue a query to validate the aggregated results.

I've added some test for the adapters itself. The Query is already covered by the existing tests.

**Comment by fooker** [comment]
Rebased to `features/drift2`

**Comment by Naicisum** [comment]
Initial testing looks good to me

**Comment by fooker** [comment]
@j-white Same for me. What about `update_switched`? Still not perfect but maybe less confusing...

**Comment by Naicisum** [comment]
`delta_switched`?

**Comment by fooker** [comment]
> `delta_switched`?

so shall it be

**Pull request merged**